_posts/2018-11-01-http-part6.markdown

```diff
@@ -0,0 +1,514 @@
+
+    fun stats(): ConnectionClientPoolStats = ConnectionClientPoolStats(connectionsEstablished.sum())
+
+    fun releaseConnection(connection: Connection) {
+        logger.debug("Releasing connection to ${connection.destination}")
+        connections.release(connection)
+        semaphore(connection.destination).release()
+    }
+
+    private fun createConnection(destination: ConnectionDestination): Connection {
+        val socket = try {
+            Socket().also { // (4)
+                it.soTimeout = socketTimeout.toMillis().toInt()
+                it.connect(destination.toInetSocketAddress(), connectionTimeout.toMillis().toInt())
+            }
+        } catch (e: SocketTimeoutException) {
+            throw ConnectionTimeoutException(destination)
+        } catch (e: Exception) {
+            throw ConnectionException("Could not connect to $destination", e)
+        }
+        val connection = Connection(socket, destination, keepAliveTimeout, checkKeepAliveInterval)
+        logger.debug("Created connection to $destination")
+        connections.addLeased(connection)
+        connectionsEstablished.increment() // (5)
+        return connection
+    }
+
+    private fun leaseConnection(destination: ConnectionDestination): Connection? {
+        logger.trace("Leasing a connection to $destination")
+        val connection = connections.lease(destination)
+        when (connection) {
+            null -> logger.trace("There was no connection to lease")
+            else -> logger.debug("Connection to $destination leased")
+        }
+        return connection
+    }
+
+
+    private fun scheduleClosingStaleConnections() { // 6
+        checkerThreadPool.scheduleAtFixedRate({
+            try {
+                connections.closeStale()
+            } catch (e: Exception) {
+                logger.error("Error while closing stale connections", e)
+            }
+        }, checkKeepAliveInterval.toMillis(), checkKeepAliveInterval.toMillis(), TimeUnit.MILLISECONDS)
+    }
+
+    override fun close() {
+        checkerThreadPool.shutdown()
+        checkerThreadPool.awaitTermination(1, TimeUnit.SECONDS)
+    }
+
+}
+```
+To retrieve a connection we have to acquire a permit (1). A permit is a semaphore that guards the pool against creating more than the limit of connections per destination. If we acquired all the permits, then the next request for a connection will wait for the max of `connectionRequestTimeout` duration (2). 
+If we successfully acquired the permit it either means that we can lease a connection or create one (3). Firstly, we try to lease a connection because we’d rather reuse a connection than create a new one. If we succeeded, we return it to the user. Otherwise, we need to create a new one. While connecting, we have to set socket and connection timeouts (4). A new connection is added to the `connections` as leased and immediately returned to the client.
```

Comment:
If we successfully acquired the permit it either means that we can lease a connection or create one ->
If we successfully acquired the permit it means that we can  either lease a connection or create one